api/client/webclient/webclient.go

```diff
@@ -48,20 +50,61 @@ func newWebClient(insecure bool, pool *x509.CertPool) *http.Client {
 	}
 }
 
-// Find fetches discovery data by connecting to the given web proxy address.
-// It is designed to fetch proxy public addresses without any inefficiencies.
-func Find(ctx context.Context, proxyAddr string, insecure bool, pool *x509.CertPool) (*PingResponse, error) {
-	clt := newWebClient(insecure, pool)
-	defer clt.CloseIdleConnections()
+// doWithFallback attempts to execute an HTTP request using https, and then
+// fall back to plain HTTP under certain, very specific circumstances.
+//  * The caller must specifically allow it via the allowPlainHTTP parameter, and
+//  * The target host must resolve to the loopback address.
+// If these conditions are not met, then the plain-HTTP fallback is not allowed,
+// and a the HTTPS failure will be considered final.
+func doWithFallback(ctx context.Context, clt *http.Client, allowPlainHTTP bool, method string, proxyAddr string, path string) (*http.Response, error) {
+	// first try https and see how that goes
+	endpoint := fmt.Sprintf("https://%s%s", proxyAddr, path)
+
+	log.Printf("Attempting %s %s", method, endpoint)
+	req, err := http.NewRequestWithContext(ctx, method, endpoint, nil)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
 
-	endpoint := fmt.Sprintf("https://%s/webapi/find", proxyAddr)
+	resp, err := clt.Do(req)
 
-	req, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, nil)
+	// If the HTTPS succeeds, return that.
+	if err == nil {
+		return resp, nil
+	}
+
+	// If we're not allowed to try plain HTTP, bail out with whatever error we have.
+	// Note that we're only allowed to try plain HTTP on the loopback address, even
+	// if the caller says its OK
+	if !(allowPlainHTTP && utils.IsLoopback(proxyAddr)) {
+		return nil, trace.Wrap(err)
+	}
+
+	// If we get to here a) the HTTPS attempt failed, and b) we're allowed to try
+	// clear-text HTTP to see if that works.
+	endpoint = fmt.Sprintf("http://%s%s", proxyAddr, path)
+	log.Printf("Falling back to plain %s %s", method, endpoint)
```

Comment:
Let's make it clear this is bad with a "WARN " prefix, message change up to you. (I'd rather not log the entire endpoint, who knows what params may be there?)

```suggestion
	log.Printf("WARN Request for %v/%v falling back to plain HTTP", req.URL.Host, req.URL.Path)
```

Reply:
Done